src/extensions/sb-ncbr/partial-charges/color.ts

```diff
@@ -0,0 +1,135 @@
+import { Bond, StructureElement, StructureProperties, Unit } from '../../../mol-model/structure';
+import { ColorTheme } from '../../../mol-theme/color';
+import { ThemeDataContext } from '../../../mol-theme/theme';
+import { Color } from '../../../mol-util/color';
+import { ParamDefinition as PD } from '../../../mol-util/param-definition';
+import { Location } from '../../../mol-model/location';
+import { SbNcbrPartialChargesPropertyProvider, isApplicable } from './property';
+import { CustomProperty } from '../../../mol-model-props/common/custom-property';
+
+const Colors = {
+    Bond: Color(0xffffff),
+    Error: Color(0x00ff00),
+    MissingCharge: Color(0xffffff),
+};
+
+export const PartialChargesThemeParams = {
+    maxAbsoluteCharge: PD.Numeric(0, undefined, {
+        label: 'Charge range',
+    }),
+    absolute: PD.Boolean(false, { isHidden: false, label: 'Use range' }),
+    showResidueCharge: PD.Boolean(true, { isHidden: false }),
+    typeId: PD.Numeric(1, undefined, { isHidden: true }),
+};
+export type PartialChargesThemeParams = typeof PartialChargesThemeParams;
+
+export function getPartialChargesThemeParams() {
+    return PD.clone(PartialChargesThemeParams);
+}
+
+function getColor(charge: number, maxCharge: number): Color {
+    const colors = {
+        negative: Color(0xff0000),
+        zero: Color(0xffffff),
+        positive: Color(0x0000ff),
+    };
+
+    if (charge === 0) return colors.zero;
+    if (charge <= -maxCharge) return colors.negative;
+    if (charge >= maxCharge) return colors.positive;
+
+    const t = maxCharge !== 0 ? Math.abs(charge) / maxCharge : 1;
+    const endColor = charge < 0 ? colors.negative : colors.positive;
+
+    return Color.interpolate(colors.zero, endColor, t);
+}
+
+export function PartialChargesColorTheme(
+    ctx: ThemeDataContext,
+    props: PD.Values<PartialChargesThemeParams>
+): ColorTheme<PartialChargesThemeParams> {
+    const model = ctx.structure?.models[0];
+    if (!model) {
+        throw new Error('No model found');
+    }
+    const data = SbNcbrPartialChargesPropertyProvider.get(model).value?.data;
+    const typeId = SbNcbrPartialChargesPropertyProvider.getParams(model).typeId.defaultValue;
+
+    // forces coloring updates
+    props.typeId = typeId;
```

Comment:
I couldn't figure out how to update the coloring when switching to a different charge set. I noticed that the coloring changes when a parameter is changed so that's why I used this.

Reply:
Color themes support a `contextHash` for that purpose, see https://github.com/molstar/molstar/blob/804117475b651da9dd4f4be86bc3a6d9738d2837/src/mol-theme/color/secondary-structure.ts#L114